File: wrench-app/src/main/java/com/izettle/wrench/applicationlist/ApplicationAdapter.java
package com.izettle.wrench.applicationlist;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.izettle.wrench.database.WrenchApplication;
import com.izettle.wrench.databinding.ApplicationListItemBinding;

import androidx.annotation.NonNull;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.DiffUtil;

public class ApplicationAdapter extends PagedListAdapter<WrenchApplication, ApplicationViewHolder> {
    private static final DiffUtil.ItemCallback<WrenchApplication> DIFF_CALLBACK = new DiffUtil.ItemCallback<WrenchApplication>() {
        @Override
        public boolean areItemsTheSame(@NonNull WrenchApplication oldApplication, @NonNull WrenchApplication newApplication) {
            return oldApplication.getId() == newApplication.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull WrenchApplication oldApplication, @NonNull WrenchApplication newApplication) {
            return oldApplication.equals(newApplication);
        }
    };

    ApplicationAdapter() {
        super(DIFF_CALLBACK);
    }

    @Override
    public ApplicationViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ApplicationViewHolder(ApplicationListItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false));
    }

    @Override
    public void onBindViewHolder(ApplicationViewHolder holder, int position) {
        WrenchApplication application = getItem(position);

        if (application != null) {
            holder.bindTo(application);
        } else {
            holder.clear();
        }
    }
}

File: wrench-app/src/main/java/com/izettle/wrench/applicationlist/ApplicationAdapter.kt
package com.izettle.wrench.applicationlist

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.DiffUtil
import com.izettle.wrench.database.WrenchApplication
import com.izettle.wrench.databinding.ApplicationListItemBinding

class ApplicationAdapter : PagedListAdapter<WrenchApplication, ApplicationViewHolder>(DIFF_CALLBACK) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ApplicationViewHolder {
        return ApplicationViewHolder(ApplicationListItemBinding.inflate(LayoutInflater.from(parent.context), parent, false))
    }

    override fun onBindViewHolder(holder: ApplicationViewHolder, position: Int) {
        val application = getItem(position)

        if (application != null) {
            holder.bindTo(application)
        } else {
            holder.clear()
        }
    }

    companion object {
        private val DIFF_CALLBACK = object : DiffUtil.ItemCallback<WrenchApplication>() {
            override fun areItemsTheSame(oldApplication: WrenchApplication, newApplication: WrenchApplication): Boolean {
                return oldApplication.id == newApplication.id
            }

            override fun areContentsTheSame(oldApplication: WrenchApplication, newApplication: WrenchApplication): Boolean {
                return oldApplication == newApplication
            }
        }
    }
}
